From c95d68b610f88a7ef65e080316108f4baa899d62 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Sun, 05 Sep 2021 20:54:03 +0300
Subject: [PATCH] [SW-0] preboot: gaudi2: clear flr reset

If FLR interrupt is received while FLR is handled by FW, FW issues
FLR reset by setting pcie2psoc bit FSM_CTRL register. However this
bit is not self clear. This patch adds support to clear pcie2psoc bit
as part of preboot init sequence.

Change-Id: I93547b7e657d943a15b34faf3caf95925ec79e12
---

diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index 022f063..318afd8 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -378,6 +378,7 @@
 __weak void preboot_read_straps(void) {}
 __weak void preboot_clear_wd_gpio(void) {}
 __weak void preboot_read_test_mode(void) {}
+__weak void preboot_reset_init(void) {}
 
 void preboot_main(void)
 {
@@ -408,6 +409,7 @@
 	spi_prot_status();
 
 	preboot_sp_sram_init();
+	preboot_reset_init();
 
 	preboot_device_id_init();
 
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 886d5e0..64640d4 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -17,6 +17,7 @@
 #include <pcie_phy_regs.h>
 #include <pcie_phy_masks.h>
 #include <pcie_wrap_regs.h>
+#include <pcie_wrap_masks.h>
 #include <pcie_dbi_regs.h>
 #include <pcie_dbi_masks.h>
 #include <pcie_aux_regs.h>
@@ -689,3 +690,22 @@
 	for (i = 0; i < cfg_ctx->blocks; i++)
 		gaudi2_init_block_instances(i, cfg_ctx);
 }
+
+static void clr_flr_reset(void)
+{
+	u32 val;
+
+	/*
+	 * If FLR reset is issued by FW, the reset
+	 * needs to be cleared by FW as the bit is not
+	 * self clear.
+	 */
+	val = regmap_readl(mmPCIE_WRAP_FLR_FSM_CTRL);
+	val &= ~(PCIE_WRAP_FLR_FSM_CTRL_FSM_PCIE2PSOC_RST_SET_MASK);
+	regmap_writel(val, mmPCIE_WRAP_FLR_FSM_CTRL);
+}
+
+void preboot_reset_init(void)
+{
+	clr_flr_reset();
+}
